ci/src/main.rs

```diff
@@ -0,0 +1,213 @@
+    if std {
+        debug("Build and test the default features ...");
+        run_cmd!(cargo build)?;
+        run_cmd!(cargo test)?;
+
+        if args.feature_matrix {
+            debug("Build and test with no features other than std ...");
+            run_cmd!(cargo build --no-default-features --features="std")?;
+            run_cmd!(cargo test --no-default-features --features="std")?;
+
+            debug("Build and test individual features along with \"std\":");
+            for feature in FEATURES.split(" ") {
+                let features = format!("std {}", feature);
+                if args.debug {
+                    eprintln!("\t feature \"{}\"", feature);
+                }
+                run_cmd!(cargo test --no-default-features --features="$features")?;
+            }
+
+            debug("Build std + no_std, to make sure they are not incompatible ...");
+            run_cmd!(cargo build --no-default-features --features="std no-std")?;
+        }
+
+        debug("Run the examples with the \"std\" feature enabled ...\n");
+        run_cmd!(cargo run --example bip32 7934c09359b234e076b9fa5a1abfd38e3dc2a9939745b7cc3c22a48d831d14bd)?;
+
+    } else {
+        debug("Build no_std, to make sure that cfg(test) doesn't hide any issues ...");
+        run_cmd!(cargo build --no-default-features --features="no-std")?;
+
+        debug("Test with no features other than \"no-std\" ...");
+        run_cmd!(cargo test --no-default-features --features="no-std")?;
+
+        debug("Build all the features excluding \"std\" ...");
+        let features = format!("no-std {}", FEATURES);
+        run_cmd!(cargo test --no-default-features --features="$features")?;
+
+        if args.feature_matrix {
+            debug("Build and test individual features along with \"no-std\" ...");
+            for feature in FEATURES.split(" ") {
+                let features = format!("no-std {}", feature);
+                run_cmd!(cargo test --no-default-features --features="$features")?;
+            }
+        }
+
+        debug("Run the examples without \"std\" and with \"no-std\"  ...");
+        run_cmd!(cargo run --no-default-features --features no-std --example bip32 7934c09359b234e076b9fa5a1abfd38e3dc2a9939745b7cc3c22a48d831d14bd)?;
+    }
+
+    if args.docs {
+        if is_stable_toolchain {
+            eprintln!("docs build cannot be doe with a stable toolchain");
+            process::exit(1);
+        }
+        debug("Building the docs ...");
+        run_cmd!(RUSTDOCFLAGS="--cfg docsrs" cargo doc --features="$FEATURES")?;
+    }
+
+    if args.fuzz {
+        debug("Running fuzz tests ...");
+        run_cmd!(cd fuzz; cargo test; ./travis-fuzz.sh)?;
+    }
+
+    if args.bench {
+        if is_stable_toolchain {
+            eprintln!("bench marks cannot be run with a stable toolchain");
+            process::exit(1);
+        }
+        debug("Running bench marks ...");
+        std::env::set_var("RUSTFLAGS", "--cfg=bench");
```

Comment:
Uh, `set_var` is asking for trouble, can we avoid it somehow?

Reply:
I couldn't get the macros to handle it in the `run_cmd!` calls, but I can have another go. Just for my learning, you say its risky because we could clobber current var and make it hard to debug?

Reply:
It's even more risky than that. Let's say in a few months someone decides to parallelize the tests unaware of this. Slap a few `thread::spawn` or perhaps even `rayon`, no problem, fearless concurrency, right? And then you spend days debugging super weird behavior when a program gets randomly called with weird env vars depending on "luck". Also you didn't unset the env var - another result of the dangerous pattern. Any code below that might be affected would break.

The doc of `cmd_lib` says `RUSTFLAGS=--cfg=bench cargo bench` should work. If it doesn't directly using `std::process::Command` for this case would be less bad than `set_var`.

Reply:
Thanks for the explanation, I did try using `cmd_lib` as you suggest but it didn't seem to work. I'll tray again.

Reply:
Actually, I remembered it's *even worse* than what I wrote before! `set_var` modifies global accessible by C code and no C libraries lock the environment using Rust-provided lock, so it means Rust code should not call any FFI unless it checked that it doesn't read environment. You can imagine this being very impractical and it led to soundness problems  in multiple crates including `std` itself. At this point people believe `set_var` should be `unsafe` and as I'm writing this, I realize even with `unsafe` it's super-impractical.

So it's NACK from me on calling `set_var` ever.